### PR TITLE
Support &method(:name) call for block_pass

### PR DESCRIPTION
**Pull request by tk0miya**
Pass `&method(:name)` to the block argument on method call is usually used, and well-known idiom in Ruby.  Therefore it is valuable to support the idiom on type check.

This gives a special treatment to it to support type checking.

### CODE REVIEW

File: lib/steep/type_construction.rb [hunk 1]
```diff
@@ -2401,7 +2401,17 @@ def synthesize(node, hint: nil, condition: false)
             if value_node
               type, constr = synthesize(value_node, hint: hint)
 
-              if hint.is_a?(AST::Types::Proc) && value_node.type == :sym
+              if hint.is_a?(AST::Types::Proc) && value_node.type == :send && value_node.children[1] == :method && type.is_a?(AST::Types::Name::Instance) && type.name.to_s == '::Method'
+                method_name = value_node.children[2].children[0]
+                if method = calculate_interface(AST::Types::Self.instance, private: true)&.methods&.[](method_name)
+                  # TODO: select appropriate method type from `method.method_types`.
```
Comment (tk0miya) [inline lib/steep/type_construction.rb:2407]: I'm unsure how to choose the appropriate method type from `method.method_types`. So I use the first one.

File: test/type_construction_test.rb [hunk 2]
```diff
@@ -6950,6 +6950,62 @@ def m: [T] () { (untyped) -> T } -> T
...
+r = x.map(&method(:my_to_s))
+      EOF
+
+      with_standard_construction(checker, source) do |construction, typing|
+        pair = construction.synthesize(source.node)
+
+        assert_typing_error typing, size: 1 do |errors|
+          assert_any!(errors) do |error|
+            assert_instance_of Diagnostic::Ruby::BlockTypeMismatch, error
+            assert /^\^\(::Integer\) -> X\(\d+\)$/ =~ error.expected.to_s
```
Comment (tk0miya) [inline test/type_construction_test.rb:6999]: I use regexp to check the BlockTypeMismatch#expected. Is there a better way?

Comment (soutaro) [reply test/type_construction_test.rb:6999]: Ah, in fact, you may want to use the newer `type_check_test.rb`.

File: lib/steep/type_construction.rb [hunk 3]
```diff
@@ -2401,7 +2401,18 @@ def synthesize(node, hint: nil, condition: false)
             if value_node
               type, constr = synthesize(value_node, hint: hint)
 
-              if hint.is_a?(AST::Types::Proc) && value_node.type == :sym
+              if hint.is_a?(AST::Types::Proc) && value_node.type == :send && value_node.children[1] == :method && AST::Builtin::Method.instance_type?(type)
+                method_name = value_node.children[2].children[0] #: Symbol
+                if method = calculate_interface(AST::Types::Self.instance, private: true)&.methods&.[](method_name)
+                  if overload = method.method_types.find {|method_type| method_type.type.params.nil? || method_type.type.params.optional? }
+                    type = AST::Types::Proc.new(
+                      type: method.method_types.first&.type || raise,
```
Comment (tk0miya) [inline lib/steep/type_construction.rb:2409]: Is it better to use `overload` here?
```suggestion
                      type: overload,
```

**Pull request merged**